File: artpipe/exporter.py
"""
ArtPipe 资产导出器 v0.3
将生成的角色数据导出为多种游戏引擎格式
零外部依赖，纯标准库实现
"""
import base64
import json
import time


class AssetExporter:
    """角色资产导出器，支持多种游戏引擎格式"""

    def __init__(self, char_data):
        self.data = char_data

    def export_all(self):
        """导出所有支持的格式"""
        exports = {}

        # SpriteSheet PNG 导出
        spritesheet = self.data.get("spritesheet", {})
        if spritesheet:
            exports["spritesheet"] = {
                "format": "png",
                "width": spritesheet.get("frame_width", 64),
                "height": spritesheet.get("frame_height", 80),
                "total_frames": spritesheet.get("total_frames", 0),
                "cols": spritesheet.get("cols", 8),
                "rows": spritesheet.get("rows", 0),
                "frame_map": spritesheet.get("frame_map", {}),
                "png_base64": spritesheet.get("png_base64", ""),
            }

        # Spine 格式导出
        skeleton = self.data.get("skeleton", {})
        animations = self.data.get("animations", {})
        if skeleton:
            exports["spine"] = {
                "format": "spine_json",
                "version": "4.1",
                "skeleton": {
                    "hash": self.data.get("id", ""),
                    "spine": "4.1.23",
                    "x": 0,
                    "y": -self.data.get("canvas_size", {}).get("height", 80),
                    "width": self.data.get("canvas_size", {}).get("width", 64),
                    "height": self.data.get("canvas_size", {}).get("height", 80),
                },
                "bones": self._format_spine_bones(skeleton),
                "slots": self._format_spine_slots(skeleton),
                "animations": self._format_spine_animations(animations),
            }

        # Unity 格式导出
        exports["unity"] = {
            "format": "unity_json",
            "name": self.data.get("id", "character"),
            "textureWidth": spritesheet.get("cols", 8) * spritesheet.get("frame_width", 64),
            "textureHeight": spritesheet.get("rows", 0) * spritesheet.get("frame_height", 80),
            "frames": self._format_unity_frames(spritesheet),
            "animations": {
                name: {
                    "frames": list(range(
                        info.get("start", 0),
                        info.get("start", 0) + info.get("count", 0)
                    )),
                    "fps": animations.get(name, {}).get("fps", 8),
                    "loop": animations.get(name, {}).get("loop", True),
                }
                for name, info in spritesheet.get("frame_map", {}).items()
            },
        }

        # Godot 格式导出
        exports["godot"] = {
            "format": "godot_tres",
            "resource_type": "SpriteFrames",
            "animations": [
                {
                    "name": name,
                    "speed": animations.get(name, {}).get("fps", 8),
                    "loop": animations.get(name, {}).get("loop", True),
                    "frames": self._format_godot_frames(name, spritesheet),
                }
                for name in spritesheet.get("frame_map", {})
            ],
        }

        return exports

    def _format_spine_bones(self, skeleton):
        """格式化 Spine 骨骼数据"""
        if not skeleton:
            return []
        bones = skeleton.get("bones", [])
        result = []
        for bone in bones:
            entry = {"name": bone["name"]}
            if bone.get("parent"):
                entry["parent"] = bone["parent"]
            entry["x"] = bone.get("x", 0)
            entry["y"] = bone.get("y", 0)
            result.append(entry)
        return result

    def _format_spine_slots(self, skeleton):
        """格式化 Spine 插槽数据"""
        if not skeleton:
            return []
        return skeleton.get("slots", [])

    def _format_spine_animations(self, animations):
        """格式化 Spine 动画数据"""
        result = {}
        for name, info in animations.items():
            result[name] = {
                "slots": {},
                "bones": {},
                "duration": info.get("frame_count", 1) / max(info.get("fps", 8), 1),
            }
        return result

    def _format_unity_frames(self, spritesheet):
        """格式化 Unity 帧数据"""
        frame_w = spritesheet.get("frame_width", 64)
        frame_h = spritesheet.get("frame_height", 80)
        cols = spritesheet.get("cols", 8)
        total = spritesheet.get("total_frames", 0)
        sheet_w = cols * frame_w

        frames = []
        for i in range(total):
            col = i % cols
            row = i // cols
            frames.append({
                "frame": {"x": col * frame_w, "y": row * frame_h, "w": frame_w, "h": frame_h},
                "rotated": False,
                "trimmed": False,
                "spriteSourceSize": {"x": 0, "y": 0, "w": frame_w, "h": frame_h},
                "sourceSize": {"w": frame_w, "h": frame_h},
            })
        return frames

    def _format_godot_frames(self, anim_name, spritesheet):
        """格式化 Godot 帧数据"""
        frame_w = spritesheet.get("frame_width", 64)
        frame_h = spritesheet.get("frame_height", 80)
        cols = spritesheet.get("cols", 8)
        frame_map = spritesheet.get("frame_map", {})
        sheet_w = cols * frame_w

        info = frame_map.get(anim_name, {})
        start = info.get("start", 0)
        count = info.get("count", 0)

        frames = []
        for i in range(start, start + count):
            col = i % cols
            row = i // cols
            frames.append({
                "x": col * frame_w,
                "y": row * frame_h,
                "width": frame_w,
                "height": frame_h,
            })
        return frames

File: docs/js/exporter.js
/**
 * ArtPipe - 多格式导出引擎
 * 支持 Unity / Spine / SpriteSheet / Godot 导出
 */

class AssetExporter {
    constructor(character, renderer) {
        this.character = character;
        this.renderer = renderer;
    }

    async export(format) {
        const c = this.character;
        const timestamp = new Date().toISOString().slice(0,10).replace(/-/g,'');
        const baseName = `artpipe_${c.config.type}_${c.style.pixelSize > 1 ? 'pixel' : 'smooth'}_${timestamp}`;

        switch (format) {
            case 'unity': return this.exportUnity(baseName);
            case 'spine': return this.exportSpine(baseName);
            case 'spritesheet': return this.exportSpriteSheet(baseName);
            case 'godot': return this.exportGodot(baseName);
        }
    }

    exportUnity(name) {
        const c = this.character;
        const animNames = Object.keys(c.animations);
        
        const meta = {
            name,
            format: 'Unity Package',
            version: '0.1.0',
            unityVersion: '2022.3',
            generatedAt: new Date().toISOString(),
            contents: {
                prefabs: [`${name}.prefab`],
                animatorController: `${name}.controller`,
                animations: animNames.map(a => `Animations/${name}_${a}.anim`),
                sprites: `Textures/${name}_Atlas.png`,
                materials: `Materials/${name}_Mat.mat`,
                script: `Scripts/${name}Controller.cs`
            }
        };

        const csCode = `// ArtPipe Auto-Generated Character Controller
using UnityEngine;

public class ${this.className(name)}Controller : MonoBehaviour
{
    private Animator animator;
    
    void Start()
    {
        animator = GetComponent<Animator>();
    }
    
    public void PlayIdle() => animator.Play("idle");
    public void PlayWalk() => animator.Play("walk");
    public void PlayRun() => animator.Play("run");
    public void PlayAttack() => animator.Play("attack");
    public void PlayHurt() => animator.Play("hurt");
    public void PlayDie() => animator.Play("die");
    
    public void SetSpeed(float speed)
    {
        animator.speed = speed;
    }
}`;

        const animatorYaml = `%YAML 1.1
%TAG !u! tag:unity3d.com,2011:
--- !u!91 &9100000
AnimatorController:
  m_Name: ${name}
  m_AnimatorParameters:
${animNames.map((a, i) => `  - m_Name: ${a}
    m_Type: 9
    m_DefaultFloat: 0`).join('
')}
  m_AnimatorLayers:
  - m_Name: Base Layer
    m_StateMachine: {fileID: 110700000}
    m_Mask: {fileID: 0}`;

        return {
            meta,
            files: {
                'controller.cs': csCode,
                'animator.controller.yaml': animatorYaml,
                'README.md': this.generateReadme(name, 'Unity')
            }
        };
    }

    exportSpine(name) {
        const c = this.character;
        const bones = c.bones;
        const anims = c.animations;

        const skeletonJson = {
            skeleton: { hash: c.seed, spine: '4.1', width: 200, height: 400 },
            bones: bones.map(b => ({
                name: b.name,
                parent: b.parent,
                x: b.x, y: -b.y,
                rotation: 0, scaleX: 1, scaleY: 1
            })),
            slots: c.parts.filter(p => p.layer > 0).map(p => ({
                name: p.name,
                bone: p.name === 'head' ? 'Head' : 
                      p.name === '左臂' ? 'L Arm' :
                      p.name === '右臂' ? 'R Arm' :
                      p.name.includes('腿') ? (p.name.includes('左') ? 'L Leg' : 'R Leg') : 'Spine',
                attachment: p.name
            })),
            skins: [{ name: 'default', attachments: {} }],
            animations: {}
        };

        // Convert animation frames to Spine timeline format
        for (const [animName, anim] of Object.entries(anims)) {
            const spineAnim = { bones: {} };
            for (const bone of bones) {
                spineAnim.bones[bone.name] = { 
                    translation: anim.frames.map((f, i) => ({
                        time: i / anim.fps,
                        x: (f.body_x || 0) * 0.5,
                        y: (f.body_y || 0) * 0.5
                    }))
                };
            }
            skeletonJson.animations[animName] = spineAnim;
        }

        const atlasStr = `${name}.png
size: 512,512
format: RGBA8888
filter: Linear,Linear
repeat: none
${c.parts.map(p => p.name).join('\n')}`;

        return {
            meta: {
                name,
                format: 'Spine Project',
                version: '4.1',
                generatedAt: new Date().toISOString(),
                contents: {
                    skeleton: `${name}.skel`,
                    atlas: `${name}.atlas`,
                    texture: `${name}.png`
                }
            },
            files: {
                'skeleton.json': JSON.stringify(skeletonJson, null, 2),
                'atlas.txt': atlasStr,
                'README.md': this.generateReadme(name, 'Spine')
            }
        };
    }

    async exportSpriteSheet(name) {
        const c = this.character;
        const anims = c.animations;
        
        const animList = [];
        for (const [animName, anim] of Object.entries(anims)) {
            // Sample keyframes (every 5th frame)
            const keyframes = [];
            for (let i = 0; i < anim.frames.length; i += 3) {
                keyframes.push({
                    frame: i,
                    time: (i / anim.fps).toFixed(3)
                });
            }
            animList.push({
                name: animName,
                fps: anim.fps,
                loop: anim.loop,
                totalFrames: anim.frames.length,
                keyframes
            });
        }

        const spriteData = {
            meta: {
                app: 'ArtPipe',
                version: '0.1.0',
                image: `${name}_spritesheet.png`,
                size: { w: 2048, h: 2048 },
                scale: 1,
                format: 'RGBA8888'
            },
            frames: animList,
            frameSize: { w: 128, h: 128 },
            animations: animList.map(a => ({
                name: a.name,
                frames: a.keyframes.map(k => k.frame),
                loop: a.loop
            }))
        };

        return {
            meta: {
                name,
                format: 'Sprite Sheet',
                generatedAt: new Date().toISOString(),
                contents: {
                    spritesheet: `${name}_spritesheet.png`,
                    data: `${name}_data.json`
                }
            },
            files: {
                'spritesheet.json': JSON.stringify(spriteData, null, 2),
                'README.md': this.generateReadme(name, 'SpriteSheet')
            }
        };
    }

    exportGodot(name) {
        const c = this.character;
        const anims = Object.keys(c.animations);

        const sceneTscn = `[gd_scene load_steps=3 format=3]

[ext_resource type="Texture2D" path="res://textures/${name}.png" id="1"]
[ext_resource type="Script" path="res://scripts/${name}_controller.gd" id="2"]

[node name="${this.className(name)}" type="CharacterBody2D"]
script = ExtResource("2")

[node name="Sprite2D" type="Sprite2D" parent="."]
texture = ExtResource("1")
hframes = ${anims.length * 4}

[node name="AnimationPlayer" type="AnimationPlayer" parent="."]
${anims.map((a, i) => `"libraries/${i}/${a}" = ExtResource("${i + 3}")`).join('\n')}`;

        const gdScript = `# ArtPipe Auto-Generated Character Controller (Godot 4)
extends CharacterBody2D

@export var speed: float = 200.0
@export var jump_velocity: float = -400.0

@onready var sprite = $Sprite2D
@onready var anim_player = $AnimationPlayer

var current_anim: String = "idle"

func _physics_process(delta: float) -> void:
    # Handle animation
    if velocity.length() > 10:
        if Input.is_action_pressed("run"):
            play_anim("run")
        else:
            play_anim("walk")
    else:
        play_anim("idle")

func play_anim(anim_name: String) -> void:
    if current_anim != anim_name:
        current_anim = anim_name
        anim_player.play(anim_name)

func attack() -> void:
    anim_player.play("attack")
    await anim_player.animation_finished
    
func take_damage() -> void:
    anim_player.play("hurt")
    await anim_player.animation_finished

func die() -> void:
    anim_player.play("die")`;

        return {
            meta: {
                name,
                format: 'Godot Scene',
                version: '4.2',
                generatedAt: new Date().toISOString(),
                contents: {
                    scene: `${name}.tscn`,
                    script: `${name}_controller.gd`,
                    texture: `textures/${name}.png`
                }
            },
            files: {
                'scene.tscn': sceneTscn,
                'controller.gd': gdScript,
                'README.md': this.generateReadme(name, 'Godot')
            }
        };
    }

    generateReadme(name, format) {
        const c = this.character;
        return `# ${name} - ArtPipe Generated Asset

Generated by [ArtPipe](https://github.com/kongshan001/art-pipe) on ${new Date().toLocaleDateString()}

## Character Info
- **Style**: ${c.style.name}
- **Type**: ${c.type.name}
- **Parts**: ${c.parts.length}
- **Bones**: ${c.bones.length}
- **Animations**: ${Object.keys(c.animations).join(', ')}
- **Export Format**: ${format}

## Usage
Import the generated files into your ${format} project.
See the included controller script for animation playback.

## License
Generated by ArtPipe MVP. Free for personal and commercial use.
`;
    }

    className(name) {
        return name.split('_').map(w => w.charAt(0).toUpperCase() + w.slice(1)).join('');
    }

    downloadFile(content, filename) {
        const blob = new Blob([content], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = filename;
        a.click();
        URL.revokeObjectURL(url);
    }
}

window.AssetExporter = AssetExporter;
